src/meta/src/hummock/compactor_manager.rs

```diff
@@ -12,34 +12,99 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
-use std::sync::Arc;
+use std::collections::HashMap;
+use std::sync::{Arc, Mutex};
+use std::time::SystemTime;
 
 use rand::Rng;
 use risingwave_hummock_sdk::HummockContextId;
 use risingwave_pb::hummock::subscribe_compact_tasks_response::Task;
-use risingwave_pb::hummock::SubscribeCompactTasksResponse;
+use risingwave_pb::hummock::{
+    CancelCompactTask, CompactTask, CompactTaskProgress, SubscribeCompactTasksResponse,
+};
 use tokio::sync::mpsc::{Receiver, Sender};
 
 use crate::MetaResult;
 
 const STREAM_BUFFER_SIZE: usize = 4;
 
 pub type CompactorManagerRef = Arc<CompactorManager>;
+type TaskId = u64;
 
 /// Wraps the stream between meta node and compactor node.
 /// Compactor node will re-establish the stream when the previous one fails.
 pub struct Compactor {
     context_id: HummockContextId,
     sender: Sender<MetaResult<SubscribeCompactTasksResponse>>,
+    // Tasks only register a heartbeat if they make progress since the last block.
+    // Else, they may have stalled even if they are still running within the compactor.
+    task_heartbeats: Mutex<HashMap<TaskId, TaskHeartbeat>>,
```

Comment:
I think task_heartbeats need to be initialized with existent assigned tasks, during meta node bootstrap.

Reply:
Otherwise if a problematic compactor no longer sends compaction heartbeat, meanwhile the meta is rebooted, then the tasks will never show in task_heartbeats.

Reply:
Yes, we should persist any tasks to meta, and remove them too. But the overhead (per-task) could be quite high.

See https://github.com/singularity-data/risingwave/blob/fb24e62fbdbd9006b61bd7971997ac72f2e5a56e/src/meta/src/cluster/mod.rs#L342

Reply:
Just as a note, the correct order of behaviour is to first add heartbeat (persist) then send the task creation request. Even if request fails, it's still fine as heartbeat timeout will elapse eventually.

Reply:
> Yes, we should persist any tasks to meta store and remove them when completed too

We've already persisted it in 
https://github.com/singularity-data/risingwave/blob/ef5de060f4d317866986df3f7dff1c21af8b04f7/src/meta/src/hummock/manager/compaction.rs#L25

It requires a meta store write per op.

Reply:
Then in the current model we may also need to bootstrap the compactor connections. But this may not be feasible. It may be better to bootstrap the existing tasks from any connection that does survive.

Reply:
Also, maybe decoupling the heartbeat insertion to time of assignment rather than time of sending may be beneficial.

Reply:
One issue with this is that to ensure state consistency, we need to perform hearbeat initiation as part of the task assignment transaction. Else, we may assign a task which for some reason never has its hearbeat initiated, thus leaving it stuck in limbo.

Reply:
Actually, if we simply assume that heartbeat initiation is reliable, this is not a problem.